File: ios/RecipeClipper/UI/Tour/WelcomeViewModel.swift
import Foundation
import Observation

/// The welcome's cards (#151), in order. `weekly` only with the `mealPlan` flag on.
enum WelcomeCard: Equatable {
    case app, clip, daily, weekly
}

/// Where the welcome goes when it is done.
enum WelcomeExit: Equatable {
    case done
    case openRecipe(Int64)
}

struct WelcomeUiState: Equatable {
    var cards: [WelcomeCard] = [.app, .clip, .daily]
    var page = 0
    /// The daily card mentions Chef mode only with its flag on.
    var chefMode = false
    /// "Try it" is opening the sample.
    var opening = false
    /// Set once the welcome is done; the screen hands it to its caller.
    var exit: WelcomeExit?

    var card: WelcomeCard { cards[page] }
    var isLast: Bool { page == cards.count - 1 }
}

/// The first-run welcome (#151): three or four cards, skippable, the last offering the sample
/// recipe or Start. Every way out (Skip, Start, Try it) marks it seen. Opened from Settings'
/// "Show the tour again" (`again`), it shows every tip once more too. Android's
/// `WelcomeViewModel`.
@MainActor
@Observable
final class WelcomeViewModel {
    private(set) var uiState: WelcomeUiState
    @ObservationIgnored private let tour: FirstRunTour
    @ObservationIgnored private let language: () -> String?
    @ObservationIgnored private(set) var seeding: Task<Void, Never>?

    init(
        tour: FirstRunTour,
        flags: FeatureFlags,
        again: Bool,
        // The UI's language picks the sample's (English if it isn't written in it).
        language: @escaping () -> String? = { Bundle.main.preferredLocalizations.first }
    ) {
        self.tour = tour
        self.language = language
        uiState = WelcomeUiState(
            cards: [.app, .clip, .daily] + (flags.isOn(.mealPlan) ? [.weekly] : []),
            chefMode: flags.isOn(.chefMode)
        )
        if again { tour.replay() }
        // The first welcome ever adds the sample, so it is waiting in Recipes whichever way out.
        let lang = language()
        seeding = Task { await tour.addSampleOnce(language: lang) }
    }

    func onNext() { uiState.page = min(uiState.page + 1, uiState.cards.count - 1) }

    func onPrevious() { uiState.page = max(uiState.page - 1, 0) }

    /// Skip or Start.
    func onDone() {
        tour.finishWelcome()
        uiState.exit = .done
    }

    @discardableResult
    func onTrySample() -> Task<Void, Never>? {
        guard !uiState.opening, uiState.exit == nil else { return nil }
        uiState.opening = true
        let lang = language()
        return Task {
            // After the first showing's own add, so the two never race to add it twice.
            await seeding?.value
            let id = await tour.sampleToOpen(language: lang)
            tour.finishWelcome()
            uiState.opening = false
            uiState.exit = id.map(WelcomeExit.openRecipe) ?? .done
        }
    }
}

File: app/src/main/java/com/example/recipeclipper/ui/tour/WelcomeViewModel.kt
package com.example.recipeclipper.ui.tour

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.recipeclipper.data.FirstRunTour
import com.example.recipeclipper.data.flags.FeatureFlags
import com.example.recipeclipper.data.flags.Flag
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.util.Locale
import javax.inject.Inject

/** The welcome's cards (#151), in order. [WEEKLY] only with the `mealPlan` flag on. */
enum class WelcomeCard { APP, CLIP, DAILY, WEEKLY }

/** Where the welcome goes when it is done. */
sealed interface WelcomeExit {
    data object Done : WelcomeExit
    data class OpenRecipe(val id: Long) : WelcomeExit
}

data class WelcomeUiState(
    val cards: List<WelcomeCard> = listOf(WelcomeCard.APP, WelcomeCard.CLIP, WelcomeCard.DAILY),
    val page: Int = 0,
    /** The daily card mentions Chef mode only with its flag on. */
    val chefMode: Boolean = false,
    /** "Try it" is opening the sample. */
    val opening: Boolean = false,
    /** Set once the welcome is done; the screen navigates, then calls `onExitHandled`. */
    val exit: WelcomeExit? = null
) {
    val card: WelcomeCard get() = cards[page]
    val isLast: Boolean get() = page == cards.lastIndex
}

/**
 * The first-run welcome (#151): three or four cards, skippable, the last offering the sample
 * recipe or Start. Every way out (Skip, Start, Try it, Back from the first card) marks it seen.
 * Opened from Settings' "Show the tour again" ([AGAIN_ARG]), it shows every tip once more too.
 */
@HiltViewModel
class WelcomeViewModel @Inject constructor(
    private val tour: FirstRunTour,
    flags: FeatureFlags,
    savedStateHandle: SavedStateHandle
) : ViewModel() {

    private val state = MutableStateFlow(
        WelcomeUiState(
            cards = listOfNotNull(
                WelcomeCard.APP, WelcomeCard.CLIP, WelcomeCard.DAILY,
                WelcomeCard.WEEKLY.takeIf { flags.isOn(Flag.MEAL_PLAN) }
            ),
            chefMode = flags.isOn(Flag.CHEF_MODE)
        )
    )
    val uiState: StateFlow<WelcomeUiState> = state.asStateFlow()

    init {
        if (savedStateHandle.get<Boolean>(AGAIN_ARG) == true) tour.replay()
        // The first welcome ever adds the sample, so it is waiting in Recipes whichever way out.
        viewModelScope.launch { tour.addSampleOnce(language()) }
    }

    fun onNext() = state.update { it.copy(page = (it.page + 1).coerceAtMost(it.cards.lastIndex)) }

    fun onPrevious() = state.update { it.copy(page = (it.page - 1).coerceAtLeast(0)) }

    /** Skip, Start, or Back from the first card. */
    fun onDone() {
        tour.finishWelcome()
        state.update { it.copy(exit = WelcomeExit.Done) }
    }

    fun onTrySample() {
        if (state.value.opening || state.value.exit != null) return
        state.update { it.copy(opening = true) }
        viewModelScope.launch {
            val id = tour.sampleToOpen(language())
            tour.finishWelcome()
            state.update { it.copy(opening = false, exit = id?.let(WelcomeExit::OpenRecipe) ?: WelcomeExit.Done) }
        }
    }

    fun onExitHandled() = state.update { it.copy(exit = null) }

    /** The UI's language picks the sample's (English if it isn't written in it). */
    private fun language(): String = Locale.getDefault().language

    companion object {
        const val AGAIN_ARG = "again"
    }
}
